lib/ai/model/agent.go

```diff
@@ -307,35 +329,50 @@ func parseJSONFromModel[T any](text string) (T, *invalidOutputError) {
 type planOutput struct {
 	Action       string `json:"action"`
 	Action_input any    `json:"action_input"`
+	Reasoning    string `json:"reasoning"`
 }
 
 // parsePlanningOutput parses the output of the model after asking it to plan it's next action
 // and returns the appropriate event type or an error.
-func parsePlanningOutput(text string) (*agentAction, *agentFinish, error) {
+func parsePlanningOutput(deltas <-chan string) (*AgentAction, *agentFinish, error) {
+	var text string
+	for delta := range deltas {
+		text += delta
```

Comment:
nit: `StringBuilder`

Reply:
This is actually slower here than a plain `string` due to us needing to perform computation on the entire string every iteration. `StringBuilder` only pays off if you do more writes than anything else and this isn't the case here. It incurs nontrivial overhead similar to an append when you call `builder.String()`